server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java

```diff
@@ -0,0 +1,207 @@
+import org.elasticsearch.cluster.metadata.IndexMetadata.APIBlock;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.cluster.service.MasterService;
+import org.elasticsearch.common.Priority;
+import org.elasticsearch.test.ESSingleNodeTestCase;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.TimeUnit;
+
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.is;
+
+public class MetadataIndexStateServiceBatchingTests extends ESSingleNodeTestCase {
+
+    /*
+     * This class is painfully clever and knows all sorts of implementation details about MetadataIndexStateService. It's hard to get
+     * around that, as we want to make sure that *batching* of opens and closes and adding index blocks will work as expected.
+     *
+     * In the case of opens that's easy enough, we just need to get hands-on enough to make sure that the open tasks are executed in a
+     * single batch -- we do that by blocking up the master service's tasks queue, then running the opens (so they are added to the queue),
+     * and then unblocking the master service again so everything processes in a single batch.
+     *
+     * But closing and adding blocks are both two-step processes where the second step is itself another cluster state update task, and we
+     * want to make sure that those batch up correctly, too. So there we run the same trick as with opens, but twice: we block the
+     * master service in order to batch up the initial tasks, then block it again, so that when the first tasks enqueue the second tasks
+     * they're doing so into a blocked master service, finally the master service is unblocked once more so that the second tasks process
+     * together in a batch.
+     */
+
+    public void testBatchOpenIndices() throws Exception {
+        final var masterService = getInstanceFromNode(ClusterService.class).getMasterService();
+
+        // create some indices
+        createIndex("test-1", client().admin().indices().prepareCreate("test-1"));
+        createIndex("test-2", client().admin().indices().prepareCreate("test-2"));
+        createIndex("test-3", client().admin().indices().prepareCreate("test-3"));
+        ensureGreen("test-1", "test-2", "test-3");
+
+        final var block1 = blockMasterService(masterService);
+        block1.call(); // wait for block
+
+        // fire off some opens
+        final var future1 = client().admin().indices().prepareOpen("test-1").execute();
+        final var future2 = client().admin().indices().prepareOpen("test-2", "test-3").execute();
+
+        // check the queue for the open-indices tasks
+        assertBusy(() -> assertThat(masterService.pendingTasks(), hasSize(3))); // two plus the blocking task itself
```

Comment:
This assertion should hold immediately, no need for an `assertBusy()` I think.

Reply:
I didn't think it would necessarily hold immediately, what with the `.execute()` methods returning `ActionFuture`s -- are there any guarantees how of how much processing they'll do before they return?